app/scripts/controllers/metametrics.js

```diff
@@ -0,0 +1,409 @@
+import { merge, omit } from 'lodash'
+import Analytics from 'analytics-node'
+import { ENVIRONMENT_TYPE_BACKGROUND } from '../lib/enums'
+
+const METAMETRICS_ANONYMOUS_ID = '0x0000000000000000'
+const BACKGROUND_PROCESS_PAGE = {
+  path: '/background-process',
+  title: 'Background Process',
+  url: '/background-process',
+}
+
+// flushAt controls how many events are sent to segment at once. Segment
+// will hold onto a queue of events until it hits this number, then it sends
+// them as a batch. This setting defaults to 20, but in development we likely
+// want to see events in real time for debugging, so this is set to 1 to disable
+// the queueing mechanism.
+const flushAt =
+  process.env.METAMASK_ENVIRONMENT === 'production' ? undefined : 1
+// flushInterval controls how frequently the queue is flushed to segment.
+// This happens regardless of the size of the queue. The default setting is
+// 10,000ms (10 seconds). This default is rather high, though thankfully
+// using the background process as our event handler means we don't have to
+// deal with short lived sessions that happen faster than the interval
+// e.g confirmations. This is set to 5,000ms (5 seconds) arbitrarily with the
+// intent of having a value less than 10 seconds.
+const flushInterval = 5000
+
+const segmentNoop = {
+  track(_, callback = () => undefined) {
+    // Need to call the callback so that environments without a segment id still
+    // resolve the promise from trackMetaMetricsEvent
+    return callback()
+  },
+  page() {
+    // noop
+  },
+  identify() {
+    // noop
+  },
+}
+
+/**
+ * Used to determine whether or not to attach a user's metametrics id
+ * to events that include on-chain data. This helps to prevent identifying
+ * a user by being able to trace their activity on etherscan/block exploring
+ */
+const trackableSendCounts = {
+  1: true,
+  10: true,
+  30: true,
+  50: true,
+  100: true,
+  250: true,
+  500: true,
+  1000: true,
+  2500: true,
+  5000: true,
+  10000: true,
+  25000: true,
+}
+
+export function sendCountIsTrackable(sendCount) {
+  return Boolean(trackableSendCounts[sendCount])
+}
+
+const isDevOrTestEnvironment = Boolean(
+  process.env.METAMASK_DEBUG || process.env.IN_TEST,
+)
+
+// This allows us to overwrite the metric destination for testing purposes
+const host = process.env.SEGMENT_HOST ?? undefined
+
+// We do not want to track events on development builds unless specifically
+// provided a SEGMENT_WRITE_KEY. This also holds true for test environments and
+// E2E, which is handled in the build process by never providing the SEGMENT_WRITE_KEY
+// when process.env.IN_TEST is truthy
+export const segment =
```

Comment:
I'd suggest that both of these get moved into the constructor. It'll be easier to test that way - then all of these `process.env` variables can be passed in as constructor options, so we don't have to mutate `process.env` in the tests.

Reply:
Also it's better to avoid stateful things being in module scope, because then we end up having shared state in between unit tests.